src/Calypso-Browser/ClyTextEditor.class.st

```diff
@@ -51,12 +51,3 @@ ClyTextEditor >> printIt [
 				do: [ '<error in printString: try ''Inspect it'' to debug>' ].
 			self afterSelectionInsertAndSelect: printString ]
 ]
-
-{ #category : #operations }
-ClyTextEditor >> undo [
```

Comment:
This override was added to reset the dirty state when all changes are cancelled by multiple cmd+z.
You can see that without this method the dirty indication does not disappear when the first edit is cancelled by cmd+z

Reply:
Yes I understood that. The problem here is that it interacts poorly with the save state, causing issues like #11649

If one accepts/saves the changes after performing an edit, then undos the edit via ctrl+z, it clears the dirty indication even though it shouldn't - thus preventing the user from saving after reverting a change.

Reply:
Just checking whether the undo stack is empty is not enough to determine whether the content is unchanged.

We interpret "dirty" as "different from the saved state".

And, "undo stack is empty" as a condition is not equivalent to "does not differ from the saved state".

Therein lies the issue with the original code. I understand the intention of it but it seems to be based on a faulty assumption

Reply:
I see.
Extra guard for #changesCancelled will prevent the issue:
```Smalltalk
ClyTextEditorToolMorph>>changesCancelled
	self pendingText = self editingText ifFalse: [ ^self ].
         ...
```

Reply:
Agreed. That makes sense.

Reply:
I would also suggest adding an additional comment for future reference